File: 746. Min Cost Climbing Stairs.go
package main

import "fmt"

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

var costCache map[int]int

func didHitCostCache(cache map[int]int, index int) bool {
	_, didFind := cache[index]
	return didFind
}

func recursiveClimbing(cost []int, index int) int {
	if index < 0 {
		return 0
	} else if didHitCostCache(costCache, index) {
		return costCache[index]
	} else if index <= 0 || index == 1 {
		return cost[index]
	}
	costCache[index] = cost[index] + min(recursiveClimbing(cost, index-1), recursiveClimbing(cost, index-2))
	return costCache[index]
}

func minCostClimbingStairs(cost []int) int {
	costCache = map[int]int{}
	return min(recursiveClimbing(cost, len(cost)-1), recursiveClimbing(cost, len(cost)-2))
}

func main() {
	cost := []int{1, 100, 1, 1, 1, 100, 1, 1, 100, 1}
	answer := minCostClimbingStairs(cost)
	fmt.Println("Answer: ", answer)
}

File: 746. Min Cost Climbing Stairs.py
from typing import List


class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        cache = dict()
        def dfs(index) -> int:
            if index < 0:
                return 0
            elif index in cache:
                return cache[index]
            elif index == 0 or index == 1:
                return cost[index]
            cache[index] = cost[index] + min(dfs(index-1),dfs(index-2))
            return cache[index]
        return min(dfs(len(cost)-1),dfs(len(cost)-2))

    def forwardMinCostClimbingStairs(self, cost: List[int]) -> int:

        cache = dict()

        def rcr(index):
            if index in cache: return cache[index]
            if index >= len(cost): return 0
            elif index == len(cost)-1 or index == len(cost)-2: return cost[index]
            cache[index] = min(cost[index] + rcr(index+1), cost[index] + rcr(index+2))
            return cache[index]
        
        return min(rcr(0),rcr(1))